RavenDB-22261: Inconsistency in Tree external API

redknightlois opened the pull request:
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-22261

### Additional description

This is a series of commits aimed at homogenizing the external API of Voron data structures in order to simplify the code and avoid edge cases that arise from inconsistent usage patterns along the codebase.

### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [x] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [ ] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing

### Testing by RavenDB QA team

- [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [ ] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

redknightlois committed:
RavenDB-22261: Refactoring in preparation for changing creation methods.

redknightlois committed:
RavenDB-22261: Accesses to trees are only able to be done through static methods.

redknightlois committed:
RavenDB-22261: Standarization of exceptions for read-only transactions

redknightlois commented:
Look at the last commit, this is more of an exploration for standarization of the logic inside Voron. I have a few other places where this approach can simplify a lot the current code. For example, as of right now I identified 10s of places where for example we assume implicitly non-nullness when in fact creation methods can return null among other issues. 

cc @ayende @arekpalinski

redknightlois committed:
RavenDB-22261: DisposableQueryable and exceptions standarization.

arekpalinski commented on src/Voron/Data/Fixed/FixedSizeTree.cs:
```diff
@@ -120,52 +120,60 @@ private void ThrowScopeAlreadyOpen()
+            new DirectAddScope(tree).Dispose();// verifying that we aren't holding a ptr out
 
+            // Setting the name of the tree has to happen before returning even if the return is null
+            // TODO: Make sure that is no longer the case by prohibiting the direct creation of trees.
```
TODO reminder

arekpalinski commented on src/Voron/Data/Fixed/FixedSizeTree.cs:
```diff
@@ -120,52 +120,60 @@ private void ThrowScopeAlreadyOpen()
+            var header = (FixedSizeTreeHeader.Embedded*)tree._parent.DirectRead(treeName);
             if (header == null)
-                return;
+                return false;
```
Do you think it's a valid scenario to _not_ read it? Shouldn't we revert the already introduced changes to `tree` object? For example we have already changed the name - `tree._treeName = ...;`

redknightlois replied:
Yes, in fact I thought exactly as you the first time. Maybe  I should add more comments. It can happen that we try to repurpose the instance but the instance does not exist / or does not have any item in it. BUT, we should still have repurposed the instance for a new tree. This is one of the big inconsistencies I have to address but I am not addressing it yet.

arekpalinski replied:
Any additional thoughts on this after spending more time on it?

arekpalinski commented on src/Voron/Data/Fixed/FixedSizeTree.cs:
```diff
@@ -120,52 +120,60 @@ private void ThrowScopeAlreadyOpen()
+                    if (tree._tx.Flags == TransactionFlags.ReadWrite && (tree._parent.State.Header.Flags & TreeFlags.FixedSizeTrees) != TreeFlags.FixedSizeTrees)
+                    {
+                        ref var state = ref tree._parent.State.Modify();
+                        state.Flags |= TreeFlags.FixedSizeTrees;
```
I have serious doubts if this is the place to handle that. It's _only_ about repurposing the instance. IMO it's tricky that we modify the state here. I'd prefer more explicit approach

redknightlois replied:
Agree, but that is the only way we can be backward compatible with already messed trees.

arekpalinski replied:
But in other places we explicitly apply `TreeFlags.FixedSizeTrees` flag

arekpalinski replied:
Why not to apply `TreeFlags.FixedSizeTrees` explicitly?

arekpalinski commented on src/Raven.Server/Documents/Indexes/MapReduce/MapReduceIndexBase.cs:
```diff
@@ -42,13 +42,7 @@ public override IDisposable InitializeIndexingWork(TransactionOperationContext i
-            if ((MapReduceWorkContext.MapPhaseTree.State.Header.Flags & TreeFlags.FixedSizeTrees) != TreeFlags.FixedSizeTrees)
-            {
-                ref var state = ref MapReduceWorkContext.MapPhaseTree.State.Modify();
-                state.Flags |= TreeFlags.FixedSizeTrees;
```
Are we good with _not_ calling this?

arekpalinski replied:
I see it was moved to `TryRepurposeInstance()` - please check my comment there.

arekpalinski commented on src/Voron/Data/Fixed/FixedSizeTree.cs:
```diff
@@ -120,52 +120,60 @@ private void ThrowScopeAlreadyOpen()
 
-            if (header->ValueSize != _valSize)
-                ThrowInvalidFixedSizeTreeSize(header);
+            return false;
```
Is it ever reachable code?

redknightlois replied:
Yes.

arekpalinski commented on src/Voron/Impl/Transaction.cs:
```diff
@@ -17,6 +17,8 @@
 using Constants = Voron.Global.Constants;
 using System.Diagnostics.CodeAnalysis;
+using static Sparrow.PortableExceptions;
+using static Voron.VoronExceptions;
```
Can we avoid static usings and handle that explicitly?

redknightlois replied:
I am not a big fan of static imports either, but code looks much better without having to explicly call `PortableExceptions.Throw<whatever>` or `VoronExceptions.ThrowIfReadOnly(...)`

If I would be allowed to make a single exception to it's use, this would be the one.

arekpalinski replied:
OK

arekpalinski commented on src/Voron/VoronExceptions.cs:
```diff
@@ -0,0 +1,34 @@
+    {
+        public static void ThrowIfReadOnly(LowLevelTransaction tx, string message = "A write transaction is required for this operation")
+        {
+            if (tx.Flags == TransactionFlags.Read)
```
Doesn't it have any performance implications? I think we always wanted to do checks like that outside to avoid method calls

redknightlois replied:
This pattern has been optimized inside the JIT now because Microsoft itself has been using this pattern internally too. Notice these are static methods, in static classes, with very simple logic.

arekpalinski replied:
OK

arekpalinski commented on src/Voron/Data/BTrees/Tree.cs:
```diff
@@ -126,6 +121,11 @@ public bool IsLeafCompressionSupported
 
         public bool HasNewPageAllocator { get; private set; }
 
+        public static Tree GetRoot(LowLevelTransaction llt, Slice name, TreeMutableState parentState)
```
Given that this is root, does it make sense to have `name` parameter? Shouldn't it always be `Constants.RootTreeNameSlice`?

arekpalinski commented on src/Voron/VoronExceptions.cs:
```diff
@@ -0,0 +1,34 @@
+{
+    public static class VoronExceptions
+    {
+        public static void ThrowIfReadOnly(LowLevelTransaction tx, string message = "A write transaction is required for this operation")
```
Why do we no longer apply `[DoesNotReturn]` to methods like that?

redknightlois replied:
Because in this case they are throw if, there is still an option for it not to fail. I will check later if I still need to add a `Throw` method marked with `[DoesNotReturn]`.

redknightlois committed:
RavenDB-22261: Standarization of Sparrow low level primitives.

redknightlois committed:
RavenDB-22261: Improvements to Encoder3Gram

redknightlois committed:
RavenDB-22261: LowLevelTransaction simplifying disposing conditions.

redknightlois committed:
RavenDB-22261: More standarization and applied review comments

redknightlois committed:
RavenDB-22261: Be explicit when null is possible when accessing structures.

maciejaszyk commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -301,16 +298,13 @@ public void Clear()
-                throw new ArgumentOutOfRangeException(nameof(from), "Cannot copy data after the end of the buffer");
-
+            ThrowIfOnDebug<InvalidOperationException>(HasValue == false, $"{nameof(HasValue)} is false.");
             EnsureIsNotBadPointer();
```
Is `EnsureIsNotBadPointer` left as a placeholder for future debugging? Since the method is empty.

redknightlois replied:
It's a conditional implementation. This is the `VALIDATE` version of it:

```
        [Conditional("VALIDATE")]
        internal readonly void EnsureIsNotBadPointer()
        {
            if (_pointer->Ptr == null)
                throw new InvalidOperationException("The inner storage pointer is not initialized. This is a defect on the implementation of the ByteStringContext class");

            if (_pointer->Key == ByteStringStorage.NullKey)
                throw new InvalidOperationException("The memory referenced has already being released. This is a dangling pointer. Check your .Release() statements and aliases in the calling code.");

            if ( this.Key != _pointer->Key)
            {
                if (this.Key >> 16 != _pointer->Key >> 16)
                    throw new InvalidOperationException("The owner context for the ByteString and the unmanaged storage are different. Make sure you havent killed the allocator and kept a reference to the ByteString outside of its scope.");

                Debug.Assert((this.Key & 0x0000000FFFFFFFF) != (_pointer->Key & 0x0000000FFFFFFFF), "(this.Key & 0x0000000FFFFFFFF) != (_pointer->Key & 0x0000000FFFFFFFF)");
                throw new InvalidOperationException("The key for the ByteString and the unmanaged storage are different. This is a dangling pointer. Check your .Release() statements and aliases in the calling code.");                                    
            }
        }
```

maciejaszyk commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -1424,14 +1391,16 @@ private SegmentInformation AllocateSegment(int size)
         {
             AllocationFailed?.Invoke();
-            throw new InvalidOperationException("Allocate gave us a segment that was already disposed.");
+            PortableExceptions.Throw<InvalidOperationException>("Allocate gave us a segment that was already disposed.");
```
```suggestion
            Throw<InvalidOperationException>("Allocate gave us a segment that was already disposed.");
```

maciejaszyk commented on src/Sparrow.Server/ByteString.cs:
```diff
@@ -1472,11 +1440,9 @@ public ByteString Skip(ByteString value, int bytesToSkip, ByteStringType type =
-            if (_disposed)
-                ThrowObjectDisposed();
-
+            PortableExceptions.ThrowIfNull(value._pointer);
```
```suggestion
            ThrowIfNull(value._pointer);
```

maciejaszyk commented on src/Sparrow/Json/UnmanagedWriteBuffer.cs:
```diff
@@ -300,9 +293,9 @@ private void AllocateNextSegment(int required, bool allowGrowth)
-            if (required > ArenaMemoryAllocator.MaxArenaSize)
-                ThrowOnAllocationSizeExceeded(required, ArenaMemoryAllocator.MaxArenaSize);
-
+            ThrowIf<InvalidOperationException>(required > ArenaMemoryAllocator.MaxArenaSize, 
+                $"Tried to allocate {new Size(required, SizeUnit.Bytes)}, which exceeds maximum allocation size of {new Size(ArenaMemoryAllocator.MaxArenaSize, SizeUnit.Bytes)}");
```
Don't we allocate to generate string exception every time even if we do not throw?

redknightlois replied:
That's a good point. I will have to look into what the JIT does on that one. Most likely case is that it will get cancelled because the underlying `Throw` of the method is marked as `[DoesNotReturn]`, but we should look into it anyways.

maciejaszyk commented on src/Voron/Data/BTrees/TreeIterator.cs:
```diff
@@ -36,19 +42,18 @@ public TreeIterator(Tree tree, LowLevelTransaction tx, bool prefetch)
         {
-            if (_disposed)
-                throw new ObjectDisposedException("TreeIterator " + _tree.Name);
+            ThrowIfDisposedOnDebug(this, "TreeIterator " + _tree.Name);
```
Do we want to throw only in debug?

arekpalinski replied:
It wasn't the case in the previous code. @redknightlois?

maciejaszyk commented on src/Voron/Data/BTrees/TreeIterator.cs:
```diff
@@ -81,14 +86,12 @@ public Slice CurrentKey
-
-                if (_currentPage == null)
-                    throw new InvalidOperationException("No current page was set");
+                ThrowIfDisposedOnDebug(this, "TreeIterator " + _tree.Name);
```
Same here.

arekpalinski replied:
@redknightlois ?

maciejaszyk commented on test/FastTests/Voron/Trees/TreeRenaming.cs:
```diff
@@ -8,6 +8,7 @@
 using Xunit;
 using Voron.Global;
 using Xunit.Abstractions;
+using Elastic.Clients.Elasticsearch;
```
Unintended?

redknightlois replied:
I hate resharper and it's automatic additions. Yes.

arekpalinski replied:
it's still here

redknightlois committed:
RavenDB-22261: Adjustments after profiling.

redknightlois committed:
RavenDB-22261: ValueReader improved allocations

redknightlois commented:
Test this in DEBUG

arekpalinski commented:
test this in debug please

arekpalinski commented:
test results in Debug - https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/119/

redknightlois committed:
RavenDB-22261: Embedded Fixed Size Trees has to be handled appropriately.

redknightlois committed:
RavenDB-22261: Removed dead code and make some code more strict.

Applied readonly modifiers where applicable
Fixed mispellings
Code that is no longer used because of replacements.

redknightlois committed:
RavenDB-22261: Implementation of exceptions using an alternative string interpolation method

redknightlois committed:
RavenDB-22181: Avoid allocating cursor list.

arekpalinski commented:
test this in debug please

arekpalinski commented on src/Corax/Querying/IndexSearcher.cs:
```diff
@@ -212,7 +212,9 @@ internal Slice EncodeAndApplyAnalyzer(in FieldMetadata binding, Analyzer analyze
     //Function used to generate Slice from query parameters.
     //We cannot dispose them before the whole query is executed because they are an integral part of IQueryMatch.
     //We know that the Slices are automatically disposed when the transaction is closed so we don't need to track them.
+#if !DEBUG
```
Why only if not debug? I'm worried it might be a potential source of odd issues

redknightlois replied:
The reason why you want it in debug is because if you dont do so, you can't debug this method as Visual Studio will trip into not initialized memory and kill the process. (It sucks big time).

arekpalinski replied:
Ouch :/ Okay I understand, although I'm still a bit concerned about having that only in RElease

arekpalinski commented on src/Sparrow/PortableExceptions.cs:
```diff
@@ -0,0 +1,806 @@
+        [InterpolatedStringHandler]
+        public readonly ref struct ConditionalThrowInterpolatedStringHandler
+        {
+            private readonly StringBuilder? _builder;
```
Can we use `#nullable enable` in this file? So we won't need to enable it globally but use a file-by-file approach

redknightlois replied:
OK, but at least for `Voron`, `Sparrow` and `Corax` expect to see in the next few month added usage of those annotations.

redknightlois commented:
I don't understand why it does not allow me to comment on your comments. So I will go one by one with quotes, but first thing first. I hate resharper and it adding back stuff I explicitly remove. 

>Why not to apply TreeFlags.FixedSizeTrees explicitly?

Because one of the important changes is to make the ability to modify the state explicit. In fact, this uncovered a bug that had no impact but it was already there. It was an inconsistency on what we stored at the level of the parent.

> Any additional thoughts on this after spending more time on it?

Yes, we are handling it explicitly on the caller now. 

> It wasn't the case in the previous code. @redknightlois?

I inspected every single call manually, there is no case where it makes sense to do that check on runtime. I still leave the check on Debug for 2 reasons. 

- It is an invariant check which I am fine with it being checked.
- I will be moving more checks that meet the same criteria while we start modifying this files. Already did that for `JsonOperationContext` not long ago.

redknightlois committed:
RavenDB-22261: Applying review comments.

arekpalinski commented:
test this in debug please

arekpalinski commented on test/FastTests/Voron/Trees/TreeRenaming.cs:
```diff
@@ -53,7 +51,7 @@ public void CanRenameTree()
         }
 
-        [Fact]
+        [RavenFact(RavenTestCategory.Voron)]
```
Since you switched to `RavenFact` please adjust - [AllTestsShouldUseRavenFactOrRavenTheoryAttributes](https://jenkins.hibernatingrhinos.com/job/PR_Tests_Debug/133/testReport/junit/SlowTests.Tests/TestsInheritanceTests/AllTestsShouldUseRavenFactOrRavenTheoryAttributes/) test

arekpalinski commented:
@redknightlois what is the status here? 

- Are you going to make any more changes / profiling etc?
- What about:
> - [x] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using QA Required tag.

? 
I believe this needs to be verified before we merge this PR CC @garayx

redknightlois commented:
We need to do a performance run to avoid regressions, correct.

arekpalinski commented:
What's the status here? What about performance testing? 

If we aim to include these changes in the next stable release we need to merge it really soon. Otherwise it will be delayed by few weeks at least.

arekpalinski commented:
@redknightlois do you think it should go to 6.0 or 6.1?

redknightlois commented:
6.1

redknightlois commented:
Under risk assessment we decided to move this into the 6.2 release. We will include this PR into the feature branch instead.
https://github.com/redknightlois/ravendb/tree/arm
